Ubuntu jammy s390x: Test failures

I am seeing the following test failures on RubyCI Ubuntu s390x version Jammy (22.04.3 LTS) on the ruby/zlib latest master branch 85637fa0fc2e5f9f1b9d8fd6652f9d81fd37f4d1 with the latest ruby/ruby master branch. Could you take a look at the failures? I think you can log in to the RubyCI Ubuntu s390x server to debug. Thank you for your help.

```
$ cat /etc/os-release | grep VERSION
VERSION_ID="22.04"
VERSION="22.04.3 LTS (Jammy Jellyfish)"
VERSION_CODENAME=jammy

$ uname -m
s390x

$ gcc --version
gcc (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Copyright (C) 2021 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

```
$ which ruby
/home/jaruga/.local/ruby-05a853c2f2-debug/bin/ruby

$ ruby -v
ruby 3.3.0dev (2023-09-11T15:25:06Z master 05a853c2f2) [s390x-linux]
```

```
$ bundle install --standalone

$ bundle list
Gems included by the bundle:
  * power_assert (2.0.3)
  * rake (13.0.6)
  * rake-compiler (1.2.5)
  * test-unit (3.6.1)
  * test-unit-ruby-core (1.0.2)
  * zlib (3.0.0)
Use `bundle info` to print more detailed information about a gem

$ bundle exec rake compile
```

```
$ ldd ./lib/zlib.so
  linux-vdso64.so.1 (0x000003ffa007e000)
  libruby.so.3.3 => /home/jaruga/.local/ruby-05a853c2f2-debug/lib/libruby.so.3.3 (0x000003ff9f900000)
  libz.so.1 => /lib/s390x-linux-gnu/libz.so.1 (0x000003ff9f800000)
  libm.so.6 => /lib/s390x-linux-gnu/libm.so.6 (0x000003ff9f700000)
  libc.so.6 => /lib/s390x-linux-gnu/libc.so.6 (0x000003ff9f500000)
  libgmp.so.10 => /lib/s390x-linux-gnu/libgmp.so.10 (0x000003ff9f400000)
  libcrypt.so.1 => /lib/s390x-linux-gnu/libcrypt.so.1 (0x000003ff9f380000)
  /lib/ld64.so.1 (0x000003ffa0000000)
```

Below is the used Zlib deb package version.

```
$ dpkg -S /lib/s390x-linux-gnu/libz.so.1
zlib1g:s390x: /lib/s390x-linux-gnu/libz.so.1
$ dpkg -s zlib1g | grep ^Version
Version: 1:1.2.11.dfsg-2ubuntu9.2
```

```
$ bundle exec rake test
Loaded suite /home/jaruga/git/ruby/zlib/bundle/ruby/3.3.0+0/gems/rake-13.0.6/lib/rake/rake_test_loader
Started
F
======================================================================================================
Failure: test_deflate_stream(TestZlib)
/home/jaruga/git/ruby/zlib/test/zlib/test_zlib.rb:1411:in `test_deflate_stream'
     1408:         deflated << chunk
     1409:       end
     1410: 
  => 1411:       assert_equal 20016, deflated.length
     1412:     end
     1413: 
     1414:     def test_gzip
<20016> expected but was
<21085>

diff:
? 2 0016
?  1 85 
?  + ??? 
======================================================================================================
P
======================================================================================================
Pending: test_gunzip_no_memory_leak(TestZlib): pended.
/home/jaruga/git/ruby/zlib/bundle/ruby/3.3.0+0/gems/test-unit-ruby-core-1.0.2/lib/core_assertions.rb:192:in `rescue in assert_no_memory_leak'
/home/jaruga/git/ruby/zlib/bundle/ruby/3.3.0+0/gems/test-unit-ruby-core-1.0.2/lib/core_assertions.rb:150:in `assert_no_memory_leak'
/home/jaruga/git/ruby/zlib/test/zlib/test_zlib.rb:1468:in `test_gunzip_no_memory_leak'
     1465:     end
     1466: 
     1467:     def test_gunzip_no_memory_leak
  => 1468:       assert_no_memory_leak(%[-rzlib], "#{<<~"{#"}", "#{<<~'};'}")
     1469:       d = Zlib.gzip("data")
     1470:       {#
     1471:         10_000.times {Zlib.gunzip(d)}
======================================================================================================
F
======================================================================================================
Failure: test_gzip(TestZlib)
/home/jaruga/git/ruby/zlib/test/zlib/test_zlib.rb:1419:in `test_gzip'
     1416:       actual[4, 4] = "\x00\x00\x00\x00" # replace mtime
     1417:       actual[9] = "\xff" # replace OS
     1418:       expected = %w[1f8b08000000000000ff4bcbcf07002165738c03000000].pack("H*")
  => 1419:       assert_equal expected, actual
     1420: 
     1421:       actual = Zlib.gzip("foo".freeze, level: 0)
     1422:       actual[4, 4] = "\x00\x00\x00\x00" # replace mtime
<"\x1F\x8B\b\x00\x00\x00\x00\x00\x00\xFFK\xCB\xCF\a\x00!es\x8C\x03\x00\x00\x00"> expected but was
<"\x1F\x8B\b\x00\x00\x00\x00\x00\x00\xFFJ\xCB\xCF\a\f\x00!es\x8C\x03\x00\x00\x00">

diff:
? �K�� !es�
?           J   
                         
?           ?   +         
======================================================================================================
F
======================================================================================================
Failure: test_deflate_chunked(TestZlibDeflate)
/home/jaruga/git/ruby/zlib/test/zlib/test_zlib.rb:66:in `test_deflate_chunked'
     63: 
     64:       final = z.finish
     65: 
  => 66:       assert_equal 7253, final.length
     67: 
     68:       chunks << final
     69:       all = chunks.join
<7253> expected but was
<8325>

diff:
? 7 253
? 83   
? ?  -
======================================================================================================
F
======================================================================================================
Failure: test_deflate_chunked_break(TestZlibDeflate)
/home/jaruga/git/ruby/zlib/test/zlib/test_zlib.rb:92:in `test_deflate_chunked_break'
     89: 
     90:       final = z.finish
     91: 
  => 92:       assert_equal 3632, final.length
     93: 
     94:       all = chunks.join
     95:       all << final
<3632> expected but was
<4702>

diff:
? 3632
? 470 
? ??? 
======================================================================================================
F
======================================================================================================
Failure: test_unused2(TestZlibGzipReader)
/home/jaruga/git/ruby/zlib/test/zlib/test_zlib.rb:968:in `test_unused2'
     965:       io = Zlib::GzipReader.new zio
     966:       assert_equal('aaaa', io.read)
     967:       unused = io.unused
  => 968:       assert_equal(24, unused.bytesize)
     969:       io.finish
     970: 
     971:       zio.pos -= unused.length
<24> expected but was
<23>

diff:
? 24
?  3
?  ?
======================================================================================================
|
Finished in 6.146844438 seconds.
------------------------------------------------------------------------------------------------------
95 tests, 511 assertions, 5 failures, 0 errors, 1 pendings, 0 omissions, 0 notifications
93.6842% passed
------------------------------------------------------------------------------------------------------
15.46 tests/s, 83.13 assertions/s
rake aborted!
Command failed with status (1)
/home/jaruga/git/ruby/zlib/bundle/ruby/3.3.0+0/gems/rake-13.0.6/exe/rake:27:in `<top (required)>'
/home/jaruga/.local/ruby-05a853c2f2-debug/bin/bundle:25:in `load'
/home/jaruga/.local/ruby-05a853c2f2-debug/bin/bundle:25:in `<main>'
Tasks: TOP => test
(See full trace by running task with --trace)
```

## ruby/ruby

I also can see the failures in ruby/ruby latest master branch `05a853c2f21f60f9e1c544c2d0709f10de453571`.

```
$ make V=1 test-all TESTS="test/zlib/test_zlib.rb -v"
...
  1) Failure:
TestZlib#test_deflate_stream [/home/jaruga/git/ruby/ruby/test/zlib/test_zlib.rb:1411]:
<20016> expected but was
<21085>.

  2) Failure:
TestZlib#test_gzip [/home/jaruga/git/ruby/ruby/test/zlib/test_zlib.rb:1419]:
<"\x1F\x8B\b\x00\x00\x00\x00\x00\x00\xFFK\xCB\xCF\a\x00!es\x8C\x03\x00\x00\x00"> expected but was
<"\x1F\x8B\b\x00\x00\x00\x00\x00\x00\xFFJ\xCB\xCF\a\f\x00!es\x8C\x03\x00\x00\x00">.

  3) Failure:
TestZlibDeflate#test_deflate_chunked [/home/jaruga/git/ruby/ruby/test/zlib/test_zlib.rb:66]:
<7253> expected but was
<8325>.

  4) Failure:
TestZlibDeflate#test_deflate_chunked_break [/home/jaruga/git/ruby/ruby/test/zlib/test_zlib.rb:92]:
<3632> expected but was
<4702>.

  5) Failure:
TestZlibGzipReader#test_unused2 [/home/jaruga/git/ruby/ruby/test/zlib/test_zlib.rb:968]:
<24> expected but was
<23>.

Finished tests in 2.985270s, 31.4879 tests/s, 236.1595 assertions/s.
94 tests, 705 assertions, 5 failures, 0 errors, 0 skips

ruby -v: ruby 3.3.0dev (2023-09-11T15:25:06Z master 05a853c2f2) [s390x-linux]
make: *** [uncommon.mk:914: yes-test-all] Error 5

...
```

--- a/.travis.yml
+++ b/.travis.yml
@@ -1,0 +1,5 @@
+dist: jammy
+language: ruby
+matrix:
+  include:
+    - arch: s390x

--- a/test/zlib/test_zlib.rb
+++ b/test/zlib/test_zlib.rb
@@ -12,6 +12,10 @@ rescue LoadError
 end
 
 if defined? Zlib
+  child_env = {}
+  child_env['DFLTCC'] = '0' if RUBY_PLATFORM =~ /s390x/
+  Zlib::CHILD_ENV = child_env.freeze
+
   class TestZlibDeflate < Test::Unit::TestCase
     def test_initialize
       z = Zlib::Deflate.new
@@ -44,59 +48,63 @@ if defined? Zlib
     end
 
     def test_deflate_chunked
-      original = ''.dup
-      chunks = []
-      r = Random.new 0
+      assert_separately([Zlib::CHILD_ENV, '-rzlib'], <<~'end;')
+        original = ''.dup
+        chunks = []
+        r = Random.new 0
 
-      z = Zlib::Deflate.new
+        z = Zlib::Deflate.new
 
-      2.times do
-        input = r.bytes(20000)
-        original << input
-        z.deflate(input) do |chunk|
-          chunks << chunk
+        2.times do
+          input = r.bytes(20000)
+          original << input
+          z.deflate(input) do |chunk|
+            chunks << chunk
+          end
         end
-      end
 
-      assert_equal [16384, 16384],
-                   chunks.map { |chunk| chunk.length }
+        assert_equal [16384, 16384],
+                     chunks.map { |chunk| chunk.length }
 
-      final = z.finish
+        final = z.finish
 
-      assert_equal 7253, final.length
+        assert_equal 7253, final.length
 
-      chunks << final
-      all = chunks.join
+        chunks << final
+         all = chunks.join
 
-      inflated = Zlib.inflate all
+        inflated = Zlib.inflate all
 
-      assert_equal original, inflated
+        assert_equal original, inflated
+      end;
     end
 
     def test_deflate_chunked_break
-      chunks = []
-      r = Random.new 0
+      assert_separately([Zlib::CHILD_ENV, '-rzlib'], <<~'end;')
+        chunks = []
+        r = Random.new 0
 
-      z = Zlib::Deflate.new
+        z = Zlib::Deflate.new
 
-      input = r.bytes(20000)
-      z.deflate(input) do |chunk|
-        chunks << chunk
-        break
-      end
+        input = r.bytes(20000)
+        z.deflate(input) do |chunk|
+          chunks << chunk
+          break
+        end
 
-      assert_equal [16384], chunks.map { |chunk| chunk.length }
+        assert_equal [16384], chunks.map { |chunk| chunk.length }
 
-      final = z.finish
+        final = z.finish
 
-      assert_equal 3632, final.length
+        assert_equal 3632, final.length
 
-      all = chunks.join
-      all << final
+        all = chunks.join
+        all << final
 
-      original = Zlib.inflate all
+        original = Zlib.inflate all
 
-      assert_equal input, original
+        assert_equal input, original
+      end;
     end
 
     def test_addstr
@@ -952,30 +960,32 @@ if defined? Zlib
     end
 
     def test_unused2
-      zio = StringIO.new
+      assert_separately([Zlib::CHILD_ENV, '-rzlib', '-rstringio'], <<~'end;')
+        zio = StringIO.new
 
-      io = Zlib::GzipWriter.new zio
-      io.write 'aaaa'
-      io.finish
+        io = Zlib::GzipWriter.new zio
+        io.write 'aaaa'
+        io.finish
 
-      io = Zlib::GzipWriter.new zio
-      io.write 'bbbb'
-      io.finish
+        io = Zlib::GzipWriter.new zio
+        io.write 'bbbb'
+        io.finish
 
-      zio.rewind
+        zio.rewind
 
-      io = Zlib::GzipReader.new zio
-      assert_equal('aaaa', io.read)
-      unused = io.unused
-      assert_equal(24, unused.bytesize)
-      io.finish
+        io = Zlib::GzipReader.new zio
+        assert_equal('aaaa', io.read)
+        unused = io.unused
+        assert_equal(24, unused.bytesize)
+        io.finish
 
-      zio.pos -= unused.length
+        zio.pos -= unused.length
 
-      io = Zlib::GzipReader.new zio
-      assert_equal('bbbb', io.read)
-      assert_equal(nil, io.unused)
-      io.finish
+        io = Zlib::GzipReader.new zio
+        assert_equal('bbbb', io.read)
+        assert_equal(nil, io.unused)
+        io.finish
+      end;
     end
 
     def test_read
@@ -1402,36 +1412,46 @@ if defined? Zlib
     end
 
     def test_deflate_stream
-      r = Random.new 0
+      assert_separately([Zlib::CHILD_ENV, '-rzlib'], <<~'end;')
+        r = Random.new 0
 
-      deflated = ''.dup
+        deflated = ''.dup
 
-      Zlib.deflate(r.bytes(20000)) do |chunk|
-        deflated << chunk
-      end
+        Zlib.deflate(r.bytes(20000)) do |chunk|
+          deflated << chunk
+        end
 
-      assert_equal 20016, deflated.length
+        assert_equal 20016, deflated.length
+      end;
     end
 
     def test_gzip
-      actual = Zlib.gzip("foo".freeze)
-      actual[4, 4] = "\x00\x00\x00\x00" # replace mtime
-      actual[9] = "\xff" # replace OS
-      expected = %w[1f8b08000000000000ff4bcbcf07002165738c03000000].pack("H*")
-      assert_equal expected, actual
+      assert_separately([Zlib::CHILD_ENV, '-rzlib'], <<~'end;')
+        actual = Zlib.gzip("foo".freeze)
+        actual[4, 4] = "\x00\x00\x00\x00" # replace mtime
+        actual[9] = "\xff" # replace OS
+        expected = %w[1f8b08000000000000ff4bcbcf07002165738c03000000].pack("H*")
+        assert_equal expected, actual
+      end;
+    end
 
+    def test_gzip_level_0
       actual = Zlib.gzip("foo".freeze, level: 0)
       actual[4, 4] = "\x00\x00\x00\x00" # replace mtime
       actual[9] = "\xff" # replace OS
       expected = %w[1f8b08000000000000ff010300fcff666f6f2165738c03000000].pack("H*")
       assert_equal expected, actual
+    end
 
+    def test_gzip_level_9
       actual = Zlib.gzip("foo".freeze, level: 9)
       actual[4, 4] = "\x00\x00\x00\x00" # replace mtime
       actual[9] = "\xff" # replace OS
       expected = %w[1f8b08000000000002ff4bcbcf07002165738c03000000].pack("H*")
       assert_equal expected, actual
+    end
 
+    def test_gzip_level_9_filtered
       actual = Zlib.gzip("foo".freeze, level: 9, strategy: Zlib::FILTERED)
       actual[4, 4] = "\x00\x00\x00\x00" # replace mtime
       actual[9] = "\xff" # replace OS

